src/header/map.rs

```diff
@@ -1057,48 +1146,51 @@ impl<T> HeaderMap<T> {
     /// valid `HeaderName`s to passed as the key (such as `String`). If they
     /// do not parse as a valid `HeaderName`, this returns an
     /// `InvalidHeaderName` error.
-    pub fn try_entry<K>(&mut self, key: K) -> Result<Entry<'_, T>, InvalidHeaderName>
+    ///
+    /// It may return `MaxSizeReached` error if size exceeds the maximum
+    /// `HeaderMap` capacity
+    pub fn try_entry<K>(&mut self, key: K) -> Result<Entry<'_, T>, Error>
     where
         K: AsHeaderName,
     {
         key.try_entry(self)
     }
 
-    fn entry2<K>(&mut self, key: K) -> Entry<'_, T>
+    fn try_entry2<K>(&mut self, key: K) -> Result<Entry<'_, T>, MaxSizeReached>
     where
         K: Hash + Into<HeaderName>,
         HeaderName: PartialEq<K>,
     {
         // Ensure that there is space in the map
-        self.reserve_one();
-
-        insert_phase_one!(
-            self,
-            key,
-            probe,
-            pos,
-            hash,
-            danger,
-            Entry::Vacant(VacantEntry {
-                map: self,
-                hash: hash,
-                key: key.into(),
-                probe: probe,
-                danger: danger,
-            }),
-            Entry::Occupied(OccupiedEntry {
-                map: self,
-                index: pos,
-                probe: probe,
-            }),
-            Entry::Vacant(VacantEntry {
-                map: self,
-                hash: hash,
-                key: key.into(),
-                probe: probe,
-                danger: danger,
-            })
-        )
+        self.try_reserve_one()?;
```

Comment:
Huh, re-reading the implementation, this surprised me. It seems like `entry()` isn't meant to allocate, unless you call `.or_insert()`. Or if it is, then `or_insert()` shouldn't allocate. Either one or the other, but probably not both... Thoughts?

(I know it was this way before, but since we're adding methods to have fallible allocation, I think it makes sense to figure this out before exposing a method that isn't needed.)